### ISSUE TITLE
Add MACOSX_DEPLOYMENT_TARGET on Darwin

### ISSUE THREAD

**Issue by sethrj** [LLNL/spack#28090]
### Summary

Define `MACOSX_DEPLOYMENT_TARGET` in the build environment on Darwin platforms based on the target OS (e.g. `bigsur`).

### Rationale

The `MACOSX_DEPLOYMENT_TARGET` environment variable provides a default `-mmacosx-version-min` argument for GCC and Clang compilers, as well as the default value of `CMAKE_OSX_DEPLOYMENT_TARGET` for CMake-based build systems. The default value for the deployment target is usually the major version (11, 10.16, ...) for CMake and Clang, but some versions of GCC specify a minor component as well (11.3), leading to numerous link warnings about inconsistent or incompatible target versions. Setting the environment variable ensures consistent versions for an install toolchain target, even when the host macOS version changes.

### Description

Add a `setup_platform_environment` to `spack/platforms/darwin.py` that sets the variable based on 

### Additional information

- Error caused by user environment's deployment version: #28014 
- Possibly related issue due to toolchain version and build environment variables: #26290
- Example of confusing warnings caused due to inconsistent gcc/clang treatment of min-macos-version: #21644
- Fallout from the GCC linker target in py-scipy: #20367

```
ld: warning: dylib (../libgfortran.dylib) was built for newer macOS version (11.1) than being linked (11.0)
```

```console
$ spack --version
0.17.0-588-1a576e732e
```

### General information

- [X] I have run `spack --version` and reported the version of Spack
- [X] I have searched the issues of this repo and believe this is not a duplicate

**Comment by sethrj** [comment]
Looks like ROOT's fork of LLVM doesn't support contemporary versions of this variable: see #28161

**Comment by trws** [comment]
What happened to #28797? I just ran into this actually killing a build due to the variable not being set while fixing up a newer version of luajit.

**Comment by sethrj** [comment]
@trws That's my bad, sorry: thought the branch was on my fork, not the mainline. I'm actually working on it this very moment: I've got one patch in place for perl (which doesn't like 11+) and am hitting an unrelated issue with `go` since I'm trying to install on an `m1` Mac for the first time.

**Comment by trws** [comment]
No worries, just happy to see you're already working on it.  It seems pre-10.5 versions didn't support `-rpath` on `ld`, and if it isn't specified in the environment `ld` just dies sometimes, I found this with luajit current so it's an easy repro.

**Comment by sethrj** [comment]
I'm seeing multiple `autoconf` packages (flex, libevent, ...) fail in linking due to "missing symbols" -- which is actually just the `configure` not adding `-Wl,-undefined -Wl,dynamic_lookup` because it has hardcoded `case .... 10.*)` in `configure`. It works "automatically" on new systems because it defaults using `${MACOSX_DEPLOYMENT_TARGET-10.0}`. Perhaps we should automatically force 10.x for autoconf packages for now? Or do you think we should add this on a package-by-package basis?

**Comment by trws** [comment]
That sounds like a good candidate for the AutotoolsPackage class to me.

**Comment by sethrj** [comment]
Done. It's going to be hard to figure out what releases are new enough *not* to need this... but honestly I just spent 10 minutes trying to find the autoconf release logs and failed. So yay!

**Comment by sethrj** [comment]
Verified that `py-scipy` installs on m1 bigsur with this changeset.

**Comment by sethrj** [comment]
Besides the ROOT and configure issues, one further place I've seen this cause a problem is gcc 8.3:
```
spack/lib/spack/env/clang/gfortran  -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk -mmacosx-version-min=11 CMakeFiles/cmTC_90190.dir/testFortranCompiler.f.o -o cmTC_90190 
    ld: -macos_version_min value malformed: '11'
```

It seems that newer versions of GCC translate `-mmacosx-version-min=11` to `-macosx_version_min 11.0.0` in the call to `ld`, but older GCC does not, and `ld` fails to parse this "version number" since it doesn't have a `.` in it. So just to be safe I will put in a PR to add a `.0` extension to the `MACOSX_DEPLOYMENT_TARGET` for newer macOS.

**Comment by adamjstewart** [comment]
I also just hit an issue with `py-scikit-build` where a version like 12 doesn't work, it needs to be the full 12.4. @sethrj did you ever open an issue for this? If not I can.

**Comment by sethrj** [comment]
> I also just hit an issue with `py-scikit-build` where a version like 12 doesn't work, it needs to be the full 12.4. @sethrj did you ever open an issue for this? If not I can.

No, I haven't seen that. Indeed the latest SDK Is I think 12.3? So surprised 12.4 works

**Comment by adamjstewart** [comment]
The latest SDK is 12.4, that's what I have on my system.

Opened a PR with a fix: #31811. Not sure why your fix in #28926 didn't work though.
